gateway/ln-gateway/src/rpc/rpc_server.rs

```diff
@@ -44,49 +39,120 @@ pub async fn run_webserver(
-        (public_routes, admin_routes)
-    } else {
-        let public_routes = Router::new()
-            .route("/set_configuration", post(set_configuration))
-            .route("/config", get(configuration))
-            // FIXME: deprecated >= 0.3.0
-            .route("/info", post(handle_post_info))
-            .route("/info", get(info));
-        let admin_routes = Router::new();
-        (public_routes, admin_routes)
-    };
+/// Extracts the Bearer token from the Authorization header of the request.
+fn extract_bearer_token(request: &Request) -> Result<String, StatusCode> {
+    let headers = request.headers();
+    let auth_header = headers.get(header::AUTHORIZATION);
+    if let Some(header_value) = auth_header {
+        let auth_str = header_value
+            .to_str()
+            .map_err(|_| StatusCode::UNAUTHORIZED)?;
+        let token = auth_str.trim_start_matches("Bearer ").to_string();
+        return Ok(token);
+    }
+
+    Err(StatusCode::UNAUTHORIZED)
+}
+
+/// Middleware to authenticate an incoming request. Routes that are
+/// authenticated with this middleware always require a Bearer token to be
+/// supplied in the Authorization header.
+async fn auth_middleware(
+    Extension(gateway): Extension<Gateway>,
+    request: Request,
+    next: Next,
+) -> Result<impl IntoResponse, StatusCode> {
+    let gateway_config = gateway.gateway_config.read().await.clone();
+    // These routes are not available unless the gateway's configuration is set.
+    let gateway_password = gateway_config.ok_or(StatusCode::NOT_FOUND)?.password;
+    authenticate(gateway_password, request, next).await
+}
+
+/// Middleware to authenticate an incoming request. Routes that are
+/// authenticated with this middleware are un-authenticated if the gateway has
+/// not yet been configured. After the gateway is configured, this middleware
+/// enforces that a Bearer token must be supplied in the Authorization header.
+async fn auth_after_config_middleware(
+    Extension(gateway): Extension<Gateway>,
+    request: Request,
+    next: Next,
+) -> Result<impl IntoResponse, StatusCode> {
+    // If the gateway's config has not been set, allow the request to continue, so
+    // that the gateway can be configured
+    let gateway_config = gateway.gateway_config.read().await.clone();
+    if gateway_config.is_none() {
+        return Ok(next.run(request).await);
+    }
+
+    // Otherwise, validate that the Bearer token matches the gateway's password
+    let gateway_password = gateway_config
+        .expect("Already validated the gateway config is not none")
+        .password;
+    authenticate(gateway_password, request, next).await
+}
+
+/// Validate that the Bearer token matches the gateway's password
+async fn authenticate(
+    gateway_password: String,
+    request: Request,
+    next: Next,
+) -> Result<axum::response::Response, StatusCode> {
+    let token = extract_bearer_token(&request)?;
+    if gateway_password == token {
```

Comment:
BTW. Do we need to use some constant time password comparison here, to prevent timing attacks?

Alternatively we should hash both sides (hash the set password in the first place and not store it anywhere in the plaintext). Can be a follow-up.

If so, can you think of any other places like this and create an issue?

Reply:
Made issue. I'm not aware of other spots where this is a problem. https://github.com/fedimint/fedimint/issues/4702

Main challenge will be doing it in a backwards compatible way.